### ISSUE TITLE
[Feature Request]: Generate source maps for CSS files

### ISSUE THREAD

**Issue by filipsobol** [rolldown/rolldown#3242]
### What problem does this feature solve?

In Rollup, we can use the `rollup-plugin-styles` plugin to process CSS files and generate source maps for them.

Rolldown can process CSS files without any additional plugins, but it doesn't generate source maps for them, even if the `output.sourcemap` option is enabled. Unfortunately, using `rollup-plugin-styles` is not an option because it fails in Rolldown with the following error:

```text
TypeError: Cannot read properties of undefined (reading 'source')
    at PluginContext.generateBundle
```

### Reproduction link or steps

https://github.com/filipsobol/rolldown-css-sourcemap-repro

1. Install dependencies.
2. Run `npm run build`.
3. Check the `dist` folder.

**Comment by nilptr** [comment]
This issue seems caused by #3240.

The `bundle` parameter received by `plugin.generateBundle(opts, bundle)` behaves differently between rolldown and rollup.
In rolldown, the `bundle` object uses a path as keys (e.g. `"dist/rolldown.js"`, `"dist/rolldown.css"`), while in rollup, it uses filenames without directories as keys (e.g. `"rollup.js"`, `"rollup.css"`).

Changing the `rolldown.config.js` to this works (`rolldown.css.map` is generated), but rolldown will write the generated files to the project root folder instead of the `dist/`.

```ts
export default defineConfig( {
  input: 'src/index.js',
  output: {
    dir: 'dist',
    file: 'rolldown.js',
    format: 'es',
    sourcemap: true,
    assetFileNames: '[name][extname]'
  },
  plugins: [
    styles( {
      mode: 'extract',
      sourceMap: true,
      minimize: true
    }),
  ],
});
```

This config does not work in rollup, because `output.dir` is not allowed for a single-file build.

**Comment by hyf0** [comment]
@nilptr Could you try `rolldown@latest`?  I think we fix the `dist` problem. `dist` would be ignored if `file` is provided.

**Comment by nilptr** [comment]
Yes, I tried with `rolldown@latest`. https://github.com/filipsobol/rolldown-css-sourcemap-repro + `rolldown@latest` + the below config solves the crash, but it will write the generated files to the project root folder.

```ts
export default defineConfig( {
  input: 'src/index.js',
  output: {
    dir: 'dist',
    file: 'rolldown.js',
    format: 'es',
    sourcemap: true,
    assetFileNames: '[name][extname]'
  },
  plugins: [
    styles( {
      mode: 'extract',
      sourceMap: true,
      minimize: true
    }),
  ],
});
```

> it will write the generated files to the project root folder

this seems expected, as `file` is provided without `'dist/'`, so `dir` is ignored.

The root cause of this issue is the discrepancy in the way rollup and rolldown handle output paths. not the old `dist` problem you mentioned.

**Comment by nilptr** [comment]
I've conducted some research on this issue. Here are some references for you:

- `bundle` keys in rollup
https://github.com/rollup/rollup/blob/15f72a8e439b254565fead053a308e07f03e4588/src/Chunk.ts#L513-L549
see line 545
- `OutputChunk.filename` in rolldown: https://github.com/nilptr/rolldown/blob/f90856ad7b4a028369dcd2627c860343910d1333/crates/rolldown/src/utils/chunk/finalize_chunks.rs#L118-L123
- https://github.com/nilptr/rolldown/blob/f90856ad7b4a028369dcd2627c860343910d1333/crates/rolldown/src/ecmascript/ecma_generator.rs#L196-L204
- how does rollup write `OutputAsset | OutputChunk` to disk: https://github.com/rollup/rollup/blob/15f72a8e439b254565fead053a308e07f03e4588/src/rollup/rollup.ts#L295-L305
- how does rolldown write `BundleOutput` to disk: https://github.com/nilptr/rolldown/blob/6bcbc401c4338fe49db917131930166071ade2db/crates/rolldown/src/bundler.rs#L96-L130
- rolldown sets a fallback value for `NormalizedBundlerOptions.dir`: https://github.com/nilptr/rolldown/blob/f90856ad7b4a028369dcd2627c860343910d1333/crates/rolldown/src/utils/normalize_options.rs#L156

**Comment by nilptr** [comment]
I would like to help on this. May I give it a try? (I might need your guidance, as I'm not aware of your long-term plans)

**Comment by hyf0** [comment]
@nilptr Sure. Thanks in advanced.

**Comment by nilptr** [comment]
Sorry, I realised I misunderstood the purpose of this issue. I put my attention on the `rollup-plugin-styles` plugin crash, which should be another topic.

#3285 is my draft for adding built-in css source map support.

**Comment by nstepien** [comment]
Looks like rolldown outputs a trailing comment like
```
//# sourceMappingURL=styles.css.map
```
in the CSS file, but `//` comments are not valid in CSS, I believe it should be `/*# sourceMappingURL=styles.css.map */` instead.

Is it possible to disable sourcemaps for CSS files?

https://github.com/adazzle/react-data-grid/issues/3736

**Comment by IWANABETHATGUY** [comment]
> Looks like rolldown outputs a trailing comment like
> 
> ```
> //# sourceMappingURL=styles.css.map
> ```
> 
> in the CSS file, but `//` comments are not valid in CSS, I believe it should be `/*# sourceMappingURL=styles.css.map */` instead.
> 
> Is it possible to disable sourcemaps for CSS files?
> 
> [adazzle/react-data-grid#3736](https://github.com/adazzle/react-data-grid/issues/3736)

Fixed in https://github.com/rolldown/rolldown/pull/3849
